utils/zerovec/src/varzerovec/owned.rs

```diff
@@ -0,0 +1,325 @@
+            );
+        }
+
+        if len == 0 {
+            // If there is no data, just construct it with the existing procedure
+            // for constructing an entire slice
+            self.entire_slice =
+                components::get_serializable_bytes(slice::from_ref(element)).unwrap();
+            return;
+        }
+
+        // The format of the encoded data is:
+        //  - four bytes of "len"
+        //  - len*4 bytes for an array of indices
+        //  - the actual data to which the indices point
+        //
+        // When inserting an element, space must be made in the `data`
+        // segment for the element, and space must be made in the `indices` segment
+        // for the new index. Note that making space in the indices segment
+        // means that the data segment needs to be shifted by 4 bytes, on top
+        // of any reshuffling that needs to be done to make space
+        //
+        // We do this in the following steps:
+        // 1. Shift the data around
+        //   1a. We move any data *after* the inserted element by the length of the element + 4
+        //       to make space for the element and one extra index
+        //   1b. We insert the element at four bytes after the spot where the data we just moved
+        //       started, again making space for that extra index
+        //   1c. We move the data from *before* the inserted element by 4 bytes to make space for the
+        //       extra index
+        // 2. Update the indices: Shift indices after the inserted element by one slot, incrementing them
+        //       by the length of the inserted element
+        // 3. Update the length
+
+        let element = element.as_unaligned();
+        let element_slice = element.as_byte_slice();
+        // the amount the vector is growing by
+        let shift = 4 + element_slice.len();
+        self.entire_slice.reserve(shift);
+        unsafe {
+            // Step 1: Shift the data around
+            {
+                let (indices, data) = self.entire_slice.split_at_mut(4 + 4 * len);
+
+                let (_len, indices) = indices.split_at_mut(4);
+                // safety: the indices at [4...4 + len * 4] form a slice of PlainOldULE<4>
+                let indices = PlainOldULE::<4>::from_byte_slice_unchecked(indices);
+
+                // Step 1a: Move the data after the inserted element by the length of
+                // the element + 4 to make space for the element and one extra index
+
+                // calculate the data insertion point as an index into `data`
+                let insertion_point = if index != len {
+                    // Assuming we're not pushing onto the end, we need to shift the tail-end elements
+                    // by `shift` to make room for one extra index and `element_slice` data
+                    let shift_start = u32::from_unaligned(&indices[index]) as usize;
+                    let shift_start_p = data.as_mut_ptr().add(shift_start);
```

Comment:
yes

Reply:
This assertion isn't correct, however, because we haven't updated the `len` of `data` though, and also it's not a vector. We're using it purely for its pointer value, we're operating on a grown vector _before_ we `.set_len()` on it (which is the correct thing, you cannot call `.set_len()` until all values have been initialized)

Reply:
Yeah.  I guess `data.capacity()` is not well-defined.  And the first case should be handled below during `data.len() - shift_start`, which should underflow if the indices are invalid for some reason.